File: rxviewmodel/src/main/java/semap/rx/viewmodel/RxViewModel.java
package semap.rx.viewmodel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModel;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.v4.util.Pair;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Observer;
import io.reactivex.Scheduler;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.ReplaySubject;

/**
 * The base class of the ViewModel.
 * @param <A> action class of the ViewModel
 * @param <S> state class of the ViewModel. It is recommended that the state class is immutable, if a property of the state is changed, a new state instance will be created.
 */
public abstract class RxViewModel<A, S> extends ViewModel {
    private static final String TAG = RxViewModel.class.getSimpleName();
    /**
     * The outputs (Observables) of the RxViewModel
     */
    private S currentState;
    private Observable<S> stateObservable;
    private Observable<ActionAndState<A, S>> actionAndStateObservable;
    private Observable<ActionAndState<A, S>> nextActionAndStateObservable;  // similar to actionAndStateObservable, but it does NOT replay
    private Observable<Throwable> errorObservable;
    private Observable<ActionAndError<A>> actionAndErrorObservable;
    private Observable<Boolean> loadingObservable;

    /**
     * The inputs (the publishEvent subjects that accepts actions) of the RxViewModel
     */
    private PublishSubject<A> concurrentActionSubject = PublishSubject.create();
    private PublishSubject<A> sequentialActionSubject = PublishSubject.create();
    private PublishSubject<A> switchMapLatestActionSubject = PublishSubject.create();

    /**
     * Internal data
     */
    protected PublishSubject<ActionAndError<A>> errorSubject = PublishSubject.create();
    private ReplaySubject<Integer> loadingCount = ReplaySubject.create();

    public RxViewModel() {
        setupRx();
    }

    private void setupRx() {
        Observable<Pair<A, StateMapper<S>>> sequential = sequentialActionSubject
                .concatMap(this::executeAndCombine);

        Observable<Pair<A, StateMapper<S>>> concurrent = concurrentActionSubject
                .flatMap(this::executeAndCombine);

        Observable<Pair<A, StateMapper<S>>> flatMapLatest = switchMapLatestActionSubject
                .switchMap(this::executeAndCombine);

        nextActionAndStateObservable = Observable.merge(sequential, concurrent, flatMapLatest)
                .observeOn(Schedulers.single())     // use trampoline to avoid race condition
                .flatMap(anm -> {
                    A action  = anm.first;
                    StateMapper<S> mapper = anm.second;
                    try {
                        S newState = mapper.map(currentState);
                        currentState = newState;
                        return Observable.just(new ActionAndState<A, S>(action, newState));
                    } catch (Throwable throwable) {
                        if (!handleError(action, throwable)) {
                            errorSubject.onNext(new ActionAndError<>(action, throwable));
                        }
                        return Observable.empty();
                    }
                })
                .observeOn(defaultScheduler())
                .share();

        actionAndStateObservable = nextActionAndStateObservable
                .replay(1)
                .autoConnect();

        currentState = createInitialState();

        stateObservable = actionAndStateObservable
                .map(ActionAndState::getState)
                .startWith(Observable.fromCallable(() -> getCurrentState()))
                .replay(1)
                .autoConnect();

        loadingObservable = loadingCount
                .scan((x, y) -> x + y)
                .map(count -> count > 0)
                .startWith(false)
                .replay(1)
                .autoConnect();

        actionAndErrorObservable = errorSubject.hide();

        errorObservable = actionAndErrorObservable
                .map(event -> event.getError());

    }

    private Observable<Pair<A, StateMapper<S>>> executeAndCombine(@NonNull A action) {
        Observable<StateMapper<S>> nonMapperObservable = Observable.just(action)
                .flatMap(a -> {
                    Observable<StateMapper<S>> stateMapperObservable = createObservable(a);
                    if (stateMapperObservable == null) {
                        stateMapperObservable = Observable.just(s -> s);
                    }

                    return stateMapperObservable
                            .defaultIfEmpty(s -> s)
                            .doOnSubscribe(__ -> {
                                if (showSpinner(a)) {
                                    loadingCount.onNext(1);
                                }
                            })
                            .doFinally(() -> {
                                if (showSpinner(a)) {
                                    loadingCount.onNext(-1);
                                }
                            })
                            .subscribeOn(defaultScheduler());
                })
                .onErrorResumeNext(throwable -> {
                    if (!handleError(action, throwable)) {
                        errorSubject.onNext(new ActionAndError<>(action, throwable));
                    }
                    return Observable.empty();
                });

        return Observable
                .combineLatest(Observable.just(action), nonMapperObservable, (a, m) -> new Pair(a, m));
    }

    /**
     * This is the method to tell the RxViewModel "how to change the current state" for a given action.
     * @param action the action needed to be executed
     * @return the StateMapper (how to change the current state)
     */
    protected abstract Observable<StateMapper<S>> createObservable(A action);

    /**
     *
     * @param action
     * @return ture if the app needs to show the spinner for the action.
     */
    protected abstract boolean showSpinner(A action);

    /**
     *
     * @return the initial state of the ViewModel
     */
    public abstract S createInitialState();

    /**
     * Default scheduler for observables
     * @return
     */
    protected Scheduler defaultScheduler() {
        return Schedulers.computation();
    }

    public <T> LiveData<T> toLiveData(Observable<T> observable) {
        Observable<T> observableWithoutError = observable
                .onErrorResumeNext(error -> {
                    if (!handleError(null, error)) {
                        errorSubject.onNext(new ActionAndError<>(null, error));
                    }
                    return Observable.empty();
                });
        return new RxLiveData<>(observableWithoutError);
    }

    /**
     * Execute the action in parallel. The View classes create Actions and call this method to run them in parallel.
     * @param action
     */
    public void executeAction(A action) {
        if (!concurrentActionSubject.hasObservers()) {
            new Handler(Looper.getMainLooper()).post(() -> execute(action, concurrentActionSubject));
        } else {
            concurrentActionSubject.onNext(action);
        }
    }

    /**
     * Execute the action in sequence. The View classes create Actions and call this method to run them in sequence.
     * @param action
     */
    public void executeActionInSequence(A action) {
        if (!concurrentActionSubject.hasObservers()) {
            new Handler(Looper.getMainLooper()).post(() -> execute(action, sequentialActionSubject));
        } else {
            sequentialActionSubject.onNext(action);
        }
    }

    // The previous action will be disposed.
    /**
     * Execute the action and dispose the previous action (if not finished).
     * The View classes create Action and call this method to run it. It will dispose the previous
     * action if the previous action is not finished.
     * @param action
     */
    public void executeActionSwitchMap(A action) {
        if (!switchMapLatestActionSubject.hasObservers()) {
            new Handler(Looper.getMainLooper()).post(() -> execute(action, switchMapLatestActionSubject));
        } else {
            switchMapLatestActionSubject.onNext(action);
        }
    }

    private void execute(A action, PublishSubject<A> publishSubject) {
        if (!publishSubject.hasObservers()) {
            // TODO Logger.error(TAG, "No observer, the action will not be executed. You need to create at least one observer of stateObservable/actionAndStateObserver or the observables from them.");
        }
        publishSubject.onNext(action);
    }

    // if the return value is true, it means the error has been handled, and it will not be in the errorSubject
    public boolean handleError(A action, Throwable error)  {
        // TODO Logger.error BBLogger.error(TAG, "Error in ViewModel " + error.getMessage());
        if (error instanceof Exception) {
            // TODO Logger.error BBLogger.error(TAG, (Exception)error);
        }
        // you can put a breakpoint here so you will get notified whenever there is an error.
        return false;
    }

    public S getCurrentState() {
        return currentState;
    }

    public Observable<S> getStateObservable() {
        return stateObservable;
    }

    public Observable<ActionAndState<A, S>> getActionAndStateObservable() {
        return getActionAndStateObservable(true);
    }

    public Observable<ActionAndState<A, S>> getActionAndStateObservable(boolean skipCurrentValue) {
        return skipCurrentValue ? nextActionAndStateObservable : actionAndStateObservable;
    }

    public Observable<ActionAndError<A>> getActionAndErrorObservable() {
        return actionAndErrorObservable;
    }

    public Observer<A> getConcurrentActionObserver() {
        return concurrentActionSubject;
    }

    public android.arch.lifecycle.Observer<A> getConcurrentActionLiveDataObserver() {
        return action -> { getConcurrentActionObserver().onNext(action);};
    }

    public Observer<A> getSequentialActionObserver() {
        return sequentialActionSubject;
    }

    public android.arch.lifecycle.Observer<A> getSequentialActionLiveDataObserver() {
        return action -> { getSequentialActionObserver().onNext(action);};
    }

    public Observer<A> getSwitchActionObserver() {
        return switchMapLatestActionSubject;
    }

    public android.arch.lifecycle.Observer<A> getSwitchActionLiveDataObserver() {
        return action -> { getSwitchActionObserver().onNext(action);};
    }

    public Observable<Throwable> getErrorObservable() {
        return errorObservable;
    }

    public Observable<Boolean> isLoadingObservable() {
        return loadingObservable;
    }

    // A util method to skip the null values.
    protected <T> Observable<T> skipNull(T object) {
        if (object == null) {
            return Observable.empty();
        }
        return Observable.just(object);
    }

    // A util method to wrap a nullable object with a Optional
    protected <T> Optional<T> toOptional(T object) {
        return new Optional<T>(object);
    }

    /**
     * Given an observable, make it showing the "spinner".
     * @param observable
     * @param <T>
     * @return
     */
    protected <T> ObservableSource<T> enableLoading(Observable<T> observable) {
        return observable
                .doOnSubscribe(__ -> {
                    increaseLoadingCount();
                })
                .doFinally(() -> {
                    decreaseLoadingCount();;
                });
    }

    /**
     * Manually increase the loading count.
     */
    public void increaseLoadingCount() {
        loadingCount.onNext(1);
    }

    /**
     * Manually decrease the loading count.
     */
    public void decreaseLoadingCount() {
        loadingCount.onNext(-1);
    }

}

File: rxviewmodel/src/main/java/semap/rx/viewmodel/RxViewModel.kt
package semap.rx.viewmodel

import androidx.lifecycle.ViewModel
import io.reactivex.rxjava3.core.Notification
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.functions.BiFunction
import io.reactivex.rxjava3.schedulers.Schedulers
import io.reactivex.rxjava3.subjects.BehaviorSubject
import io.reactivex.rxjava3.subjects.PublishSubject
import io.reactivex.rxjava3.subjects.ReplaySubject
import io.reactivex.rxjava3.subjects.UnicastSubject
import semap.rx.viewmodel.ActionExecutionMode.*

abstract class RxViewModel<A, S>: ViewModel() {

    /**
     * The outputs (Observables) of the RxViewModel
     */
    val currentState: S
        get() = stateBehaviorSubject.value ?: createInitialState()

    /**
     * Observable of State, it will replay
     */
    val stateObservable: Observable<S> by lazy {
        this.wrapperObservable
                .doOnNext {
                    if (it.isComplete) actionOnCompleteSubject.onNext(it.actionAndState)
                    else {
                        actionOnNextSubject.onNext(it.actionAndState)
                    }
                }
                .filter { !it.isComplete }
                .map { it.actionAndState.state }
                .startWith(Observable.fromCallable { createInitialState() })
                .replay(1)
                .autoConnect()
    }

    /**
     * Observable of loading, it will replay
     */
    val loadingObservable: Observable<Boolean> by lazy {
        loadingCount
                .observeOn(defaultScheduler())
                .map { it.second }
                .toBooleanObservable()
    }

    /**
     * Observable of ActionAndState, it does NOT replay
     */
    val actionOnNextObservable: Observable<ActionAndState<A, S>> by lazy {
        this.actionOnNextSubject
                .observeOn(defaultScheduler())
                .withLatestFrom(stateObservable,
                        BiFunction<ActionAndState<A, S>, S, ActionAndState<A, S>> { ans, _ -> ans })
    }

    /**
     * Observable of ActionAndState, it does NOT replay
     */
    val actionOnCompleteObservable: Observable<ActionAndState<A, S>> by lazy {
        this.actionOnCompleteSubject
                .observeOn(defaultScheduler())
                .withLatestFrom(stateObservable,
                        BiFunction<ActionAndState<A, S>, S, ActionAndState<A, S>> { ans, _ -> ans })
    }

    /**
     * Observable of ActionAndState, it does NOT replay
     */
    val actionErrorObservable: Observable<ActionAndError<A>> by lazy {
        errorSubject
                .observeOn(defaultScheduler())
                .withLatestFrom(stateObservable,
                        BiFunction<ActionAndError<A>, S, ActionAndError<A>> { ans, _ -> ans })
    }

    /**
     * Observable of Error, it does NOT replay
     */
    val errorObservable: Observable<Throwable> by lazy {
        this.actionErrorObservable
                .map { (_, error) -> error }
    }

    /**
     * The inputs (the publishEvent subjects that accepts actions) of the RxViewModel
     */
    private val concatEagerActionSubject = UnicastSubject.create<DeferrableAction>().toSerialized()
    private val concurrentActionSubject = UnicastSubject.create<A>().toSerialized()
    private val sequentialActionSubject = UnicastSubject.create<List<A>>().toSerialized()
    private val switchMapLatestActionSubject = UnicastSubject.create<A>().toSerialized()
    private val deferredActionSubject = PublishSubject.create<A>().toSerialized()

    /**
     * Internal data
     */
    private val stateBehaviorSubject = BehaviorSubject.create<S>()
    private val errorSubject = PublishSubject.create<ActionAndError<A>>().toSerialized()
    private val loadingCount = ReplaySubject.create<Pair<A, Int>>().toSerialized()
    private val actionOnNextSubject = PublishSubject.create<ActionAndState<A, S>>().toSerialized()
    private val actionOnCompleteSubject = PublishSubject.create<ActionAndState<A, S>>().toSerialized()
    private val wrapperObservable: Observable<AnsWrapper> by lazy {
        val scheduler = defaultScheduler()
        val sequential = sequentialActionSubject
                .observeOn(scheduler)
                .concatMap { list ->
                    Observable.fromIterable(list)
                            .concatMap { executeAndCombine(it, true) }
                            .onErrorResumeNext { Observable.empty<AnsWrapper>() }
                }
        val concurrent = concurrentActionSubject
                .observeOn(scheduler)
                .flatMap { executeAndCombine(it) }
        val concatEager = concatEagerActionSubject
                .observeOn(scheduler)
                .concatMapEager { executeAndCombine(it.action, throwError = false, deferredAction = it.isDeferred) }
        val flatMapLatest = switchMapLatestActionSubject
                .observeOn(scheduler)
                .switchMap { executeAndCombine(it) }
        val deferred = deferredActionSubject
                .observeOn(scheduler)
                .concatMap { executeAndCombine(it) }

        Observable
                .merge(listOf(sequential, concatEager, concurrent, flatMapLatest, deferred))
                .publish()
                .autoConnect()
    }

    /**
     *
     * @return the initial state of the ViewModel
     */
    abstract fun createInitialState(): S

    /**
     * This is the method to tell the RxViewModel "how to change the current state" for a given action.
     * @param action the action needed to be executed
     * @return the Observable of Reducers (how to change the current state). If it returns null, or an empty observable,
     * it means the given action does not change the state.
     */
    abstract fun createReducerObservable(action: A): Observable<Reducer<S>>?

    /**
     *
     * @param action
     * @return true if the app needs to show the spinner for the action.
     */
    open fun showSpinner(action: A): Boolean {
        return false
    }

    /**
     * Default scheduler for observables
     * @return
     */
    open fun defaultScheduler(): Scheduler {
        return Schedulers.computation()
    }

    // if the return value is true, it means the error has been handled, and it will not be in the errorSubject
    open fun handleError(action: A?, error: Throwable): Boolean {
        // you can put a breakpoint here so you will get notified whenever there is an error.
        return false
    }

    fun <T> toLiveData(observable: Observable<T>): RxLiveData<T> {
        val observableWithoutError = observable
                .onErrorResumeNext { error: Throwable ->
                    if (!handleError(null, error)) {
                        errorSubject.onNext(ActionAndError(null, error))
                    }
                    Observable.empty<T>()
                }
        return RxLiveData(observableWithoutError)
    }

    /**
     * Execute an action.
     * @param action
     */
    inline fun execute(action: A) {
        execute(action, ParallelDefault)
    }

    /**
     * Execute an action with a specific executionMode.
     * @param action
     * @param mode
     */
    fun execute(action: A, mode: ActionExecutionMode) {
        when(mode) {
            Parallel ->  enqueueInParallel(action, stateMapInOrder = false, deferred = false)
            ParallelDefer -> enqueueInParallel(action, stateMapInOrder = true, deferred = true)
            Sequence -> enqueueInSequence(action)
            SwitchMap -> enqueueWithSwitchMap(action)
            else -> enqueueInParallel(action, stateMapInOrder = true, deferred = false)
        }
    }

    /**
     * Execute the actions in sequence. The View classes create Actions and call this method to run them in sequence.
     * If an action fails, the rest of the actions will not be executed.
     * @param actions
     */
    fun executeInSequence(vararg actions: A) {
        sequentialActionSubject.onNext(actions.asList())
    }

    fun <T> Observable<T>.asLiveData(): RxLiveData<T> {
        return this.asLiveData(this@RxViewModel)
    }

    fun <T: A> actionOnNextObservable(clazz: Class<T>): Observable<T> {
        return actionOnNextObservable { clazz.isInstance(it) }
                .cast(clazz)
    }

    fun actionOnNextObservable(predicate: (A) -> Boolean): Observable<A> {
        return this.actionOnNextObservable
                .map { it.action }
                .filter { predicate(it) }
    }

    fun <T: A, R> actionOnNextObservable(clazz: Class<T>, mapper: (S) -> R?): Observable<R> {
        return actionOnNextObservable({ clazz.isInstance(it) } , mapper)
    }

    fun <R> actionOnNextObservable(predicate: (A) -> Boolean, mapper: (S) -> R?): Observable<R> {
        return this.actionOnNextObservable
                .filter { predicate(it.action) }
                .skipNull { mapper.invoke(it.state) }
    }

    fun <T: A> actionOnCompleteObservable(clazz: Class<T>): Observable<T> {
        return actionOnCompleteObservable { clazz.isInstance(it) }
                .cast(clazz)
    }

    fun actionOnCompleteObservable(predicate: (A) -> Boolean): Observable<A> {
        return this.actionOnCompleteObservable
                .map { it.action }
                .filter { predicate(it) }
    }

    fun <T: A, R> actionOnCompleteObservable(clazz: Class<T>, mapper: (S) -> R?): Observable<R> {
        return actionOnCompleteObservable({ clazz.isInstance(it) }, mapper)
    }

    fun <R> actionOnCompleteObservable(predicate: (A) -> Boolean, mapper: (S) -> R?): Observable<R> {
        return this.actionOnCompleteObservable
                .filter { predicate(it.action) }
                .skipNull { mapper.invoke(it.state) }
    }

    fun <T: A> actionErrorObservable(clazz: Class<T>): Observable<Throwable> {
        return actionErrorObservable { clazz.isInstance(it) }
    }

    fun actionErrorObservable(predicate: (A) -> Boolean): Observable<Throwable> {
        return this.actionErrorObservable
                .filter { it.action?.let { a -> predicate(a) } ?: false }
                .map { it.error }
    }

    fun <T: A> loadingObservable(clazz: Class<T>): Observable<Boolean> {
        return loadingObservable { clazz.isInstance(it)  }
    }

    fun loadingObservable(predicate: (A) -> Boolean): Observable<Boolean> {
        return loadingCount
                .observeOn(defaultScheduler())
                .filter { predicate(it.first) }
                .map { it.second }
                .toBooleanObservable()
    }

    private fun Observable<Int>.toBooleanObservable() : Observable<Boolean> {
        return this.scan { x, y -> x + y }
                .map { count -> count > 0 }
                .startWithItem(false)
                .distinctUntilChanged()
                .withLatestFrom(stateObservable,
                        BiFunction<Boolean, S, Boolean> { b, _ -> b })
                .replay(1)
                .autoConnect()
    }

    private fun executeAndCombine(action: A, throwError: Boolean = false, deferredAction: Boolean = false): Observable<AnsWrapper> {
        return Observable.combineLatest(
                    Observable.just(action),
                    if (deferredAction) Observable.empty() else toReducerObservable(action),
                    BiFunction<A, Reducer<S>, Pair<A, Reducer<S>>> { a, m -> Pair(a, m) })
                .observeOn(Schedulers.single())     // use Schedulers.single() to avoid race condition
                .map { ActionAndState(it.first, it.second.invoke(currentState)) }
                .doOnNext { stateBehaviorSubject.onNext(it.state) }
                .observeOn(defaultScheduler())
                .map { AnsWrapper(it) }
                .materialize()
                .doOnNext { if (deferredAction && it.isOnComplete) deferredActionSubject.onNext(action) }
                .filter { !deferredAction }
                .doOnNext {
                    it.error?.let {throwable ->
                        if (!handleError(action, throwable)) {
                            errorSubject.onNext(ActionAndError(action, throwable))
                        }
                    }
                }
                .flatMap {
                    if (it.isOnComplete) {
                        Observable.just(
                                Notification.createOnNext(AnsWrapper(ActionAndState(action, currentState), true)),
                                Notification.createOnComplete<AnsWrapper>()
                        )
                    }  else {
                        Observable.just(it)
                    }
                }
                .dematerialize { it }
                .onErrorResumeNext { throwable: Throwable ->
                    if (throwError)
                        throw throwable
                    else
                        Observable.empty<AnsWrapper>()
                }
                .doOnSubscribe {
                    if (!deferredAction && showSpinner(action)) {
                        loadingCount.onNext(Pair(action, 1))
                    }
                }
                .doFinally {
                    if (!deferredAction && showSpinner(action)) {
                        loadingCount.onNext(Pair(action, -1))
                    }
                }
                .subscribeOn(defaultScheduler())
    }

    private fun enqueueInParallel(action: A, stateMapInOrder: Boolean, deferred: Boolean) {
        if (stateMapInOrder) {
            concatEagerActionSubject.onNext(DeferrableAction(action, deferred))
        } else {
            concurrentActionSubject.onNext(action)
        }
    }

    private fun enqueueInSequence(action: A) {
        sequentialActionSubject.onNext(listOf(action))
    }

    private fun enqueueWithSwitchMap(action: A) {
        switchMapLatestActionSubject.onNext(action)
    }

    private fun toReducerObservable(action: A): Observable<Reducer<S>> {
        return Observable.just(action)
                .flatMap { a ->
                    val stateMapperObservable: Observable<Reducer<S>> =
                            createReducerObservable(a) ?: Observable.just { s -> s }

                    stateMapperObservable
                            .defaultIfEmpty { s -> s }
                }
    }

    private inner class AnsWrapper(val actionAndState: ActionAndState<A, S>, val isComplete: Boolean = false)
    private inner class DeferrableAction(val action: A, val isDeferred: Boolean = false)
}